rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+primitive cache has unique kernels and the kernel cache may have unique kernels
+too. This means that memory consumption can be higher that it is now
+
+##### Option 2: Provide API to Control Kernel Cache Capacity
+
+This option suggests adding an API to control kernel cache capacity similar to
+the one for the primitive cache.
+
+Pros:
+* Provides more control over the kernel cache and therefore more flexibility to
+the users. Those users who heavily rely on dynamic shapes can balance between
+primitive cache capacity and the kernel cache capacity
+* User can fully control memory consumption
+
+Cons:
+* Introducing yet another *cache capacity* definition that should be specified.
+Also, we will have to explain to the users the difference between kernel cache and primitive cache capacities as well as how they interoperate with each other
+
+###### C API
+
+```c
+dnnl_status_t dnnl_get_kernel_cache_capacity(int *capacity);
+dnnl_status_t dnnl_set_kernel_cache_capacity(int capacity);
+```
+
+###### C++ API
+
+```cpp
+int get_primitive_cache_capacity();
+void set_primitive_cache_capacity(int capacity);
+```
+
+###### Environment Variable
+
+`ONEDNN_PRIMITIVE_CACHE_CAPACITY=<number>`, where `number` is the kernel cache
+capacity. The default value is 1024. When `number` is 0 the kernel cache is
+disabled.
+
+###### Build Time Control
+
+The kernel cache can be enabled or disabled at the build time using a CMake option
+`ONEDNN_ENABLE_KERNEL_CACHE`.
+
+##### Options Comparison
+
+Both options assume that there will be a reasonable default capacity for the
+kernel cache. Also, both options allow users to control the caches capacity
+therefore, the only difference between the options is in the flexibility.
+The second option is in fact the first one with advanced controlling
+capabilities.
+
+Given that the first option will address the problem with dynamic shapes
+related creation overhead the proposal is to start with the it and add the
+API from the option two if we see any need in it.
+
+### Option 2: Extending Runtime Dimensions Support
+
+This option suggests extending the existing runtime dimensions support for
+relevant primitives.
+
+Pros:
+* Using the existing mechanism, the primitive cache will automatically work
+for primitives created with runtime dimensions
+* Unlike the kernel cache, with this approach the primitive and primitive descriptor
+is not re-created therefore there is no related overhead
+
+Cons:
+* The users have to modify their code to enable runtime dimensions
+* Some primitives will require to generate multiple kernels so that the
+primitives can be used with runtime dimensions. It may not be always feasible to
+generate a reasonable number of kernels
+* For some kernels it may not be possible to stay performant and support runtime
+dimensions for all dimensions. This means that if user uses incorrect combination
+of runtime dimensions they may get suboptimal performance
```

Comment:
If considering primitive and primitive desc overhead reduction as well, for example on CPU, will performance of runtime dimension primitive get comparable? As I know, currently there are some typical usages in framework side in which case the certain dimension will be dynamic while leave other dimensions static.

Reply:
> If considering primitive and primitive desc overhead reduction as well, for example on CPU, will performance of runtime dimension primitive get comparable?

I think we can't assume that in general because we can easily fall back to the reference implementation that will be very slow.

> As I know, currently there are some typical usages in framework side in which case the certain dimension will be dynamic while leave other dimensions static.

Like I said in the RFC in the conclusion paragraph, the runtime dimensions feature is good for optimizing specific cases just like the one you mentioned, when you need to support only 1 dimensions that can be runtime.

Just want to reemphasize that the proposed options are not mutually exclusive and it is possible to have a kernel cache while enabling the runtime dimensions feature for specific cases.
The RFC is about proposing a general direction to invest in and the kernel cache seems to be a good way to get better performance overall.